lapack/gonum/dlatdf.go

```diff
@@ -0,0 +1,161 @@
+package gonum
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas/blas64"
+	"gonum.org/v1/gonum/lapack"
+)
+
+// Contributors: Bo Kagstrom and Lars Westin,
+// Generalized Schur Methods with Condition Estimators for
+// Solving the Generalized Sylvester Equation, IEEE Transactions
+// on Automatic Control, Vol. 34, No. 7, July 1989, pp 745-751.
+
+// Dlatdf uses the LU factorization of the n×n matrix Z computed by Dgetc2 and
+// computes a contribution to the reciprocal Dif-estimate by solving
+//  Z * x = b for x
+// and choosing the rhs b such that the norm of x is as large as possible.
+// On entry rhs = b holds the contribution from earlier solved sub-systems, and on return rhs = x.
+//
+// The factorization of Z returned by Dgetc2 has the form
+//  Z = P*L*U*Q,
+// where P and Q are permutation matrices. L is lower triangular with
+// unit diagonal elements and U is upper triangular.
+//
+// On entry ipiv and jpiv are n length slices. On exit the pivot indices where row i has
+// been interchanged with ipiv[i] and column j interchanged with column jpiv[j].
+//
+// It must hold that n <= len(ipiv), n <= len(jpiv). rhs must have stride ldz and n number of accesible elements.
+// Dlatdf implicitly expects z matrix to be at most an 8×8 matrix. This is fulfilled by it's only caller, Dtgsy2.
+//
+// ijob info:
+// ijob==2: First compute an approximative null-vector e of Z using DGECON,
+// e is normalized and solve for Zx = +-e - f with the sign giving the greater value
+// of 2-norm(x). About 5 times as expensive as Default.
+// ijob!=2: Local look ahead strategy where all entries of the rhs. b is chosen as either +1 or -1 (Default).
+//
+// rdsum info:
+// On entry, the sum of squares of computed contributions to the Dif-estimate
+// under computation by Dtgsyl, where the scaling factor rdscal (see below)
+// has been factored out. On exit, the corresponding sum of squares updated
+// with the contributions from the current sub-system. If trans = 'T' rdsum is not touched.
+// NOTE: rdsum only makes sense when Dtgsy2 is called by STGSYL.
+//
+// rdscal info:
+// On entry, scaling factor used to prevent overflow in rdsum. On exit, rdscal is
+// updated w.r.t. the current contributions in rdsum. If trans = 'T', rdscal is not touched.
+// NOTE: rdscal only makes sense when Dtgsy2 is called by Dtgsyl.
+//
+// Dlatdf is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlatdf(ijob, n int, z []float64, ldz int, rhs []float64, rdsum, rdscal float64, ipiv, jpiv []int) (sum, scale float64) {
+	switch {
+	case n < 0:
+		panic(nLT0)
+	case ldz < max(1, n):
+		panic(badLdZ)
+	case len(rhs) < (n-1)*ldz:
+		// rhs is expected to have stride ldz
```

Comment:
The Fortran code might make it look like that but `rhs` is just a column n-vector represented by a continuous array (increment of 1). For the Dlaswp call the Fortran code needs to represent it as a matrix and so they need a stride even though it's actually not used (there is no 2. column to which stride to). They use ldz as a placeholder, they could use n just as well. But since Gonum is row-major, our code should use 1 as the stride for rhs and xm.